zokrates_core/src/static_analysis/uint_optimizer.rs

```diff
@@ -366,35 +367,25 @@ impl<'ast, T: Field> Folder<'ast, T> for UintOptimizer<'ast, T> {
                 // reduce both terms
                 let e = self.fold_uint_expression(e);
 
-                let e_max: u128 = e
-                    .metadata
-                    .clone()
-                    .unwrap()
-                    .max
-                    .to_dec_string()
-                    .parse()
-                    .unwrap();
+                let e_max: num_bigint::BigUint = e.metadata.as_ref().unwrap().max.to_biguint();
+                let max = e_max
+                    .shl(by as usize)
+                    .bitand(&(2_u128.pow(range as u32) - 1).into());
 
-                let max = T::from((e_max << by) & (2_u128.pow(range as u32) - 1));
+                let max = T::try_from(max).unwrap();
```

Comment:
Line 380 we reduce `e`, therefore I suspect `max` is `2**(bitwidth - by) - 1` no matter what `e_max` is before reducing? If this reasoning is correct, it would be clearer to reduce first in this match arm

Reply:
We can also change to *not* always reducing `e`, in which case whether we have to reduce it or not depends of `e_max`: we have to reduce if `(e_max << by) > p - 1` which I think corresponds to line 375 panicking